google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java

```diff
@@ -502,7 +506,40 @@ public boolean delete(BlobId blob) {
 
   @Override
   public Blob compose(ComposeRequest composeRequest) {
-    return todo();
+    final Map<StorageRpc.Option, ?> optionsMap =
+        StorageImpl.optionMap(Iterables.toArray(composeRequest.getTargetOptions(), Option.class));
+    GrpcCallContext grpcCallContext = GrpcRequestMetadataSupport.create(optionsMap);
+    ComposeObjectRequest.Builder composeObjectReqBuilder = ComposeObjectRequest.newBuilder();
+    final List<SourceObject> sources =
+        Lists.newArrayListWithCapacity(composeRequest.getSourceBlobs().size());
+    for (ComposeRequest.SourceBlob sourceBlob : composeRequest.getSourceBlobs()) {
+      sources.add(
+          SourceObject.newBuilder()
+              .setName(sourceBlob.getName())
+              .setGeneration(sourceBlob.getGeneration())
+              .build());
+    }
+    composeObjectReqBuilder.addAllSourceObjects(sources);
+    final Object target = codecs.blobInfo().encode(composeRequest.getTarget());
+    composeObjectReqBuilder.setDestination(target);
+    ifNonNull(
+        (Long) optionsMap.get(StorageRpc.Option.IF_GENERATION_MATCH),
+        composeObjectReqBuilder::setIfGenerationMatch);
+    ifNonNull(
+        (Long) optionsMap.get(StorageRpc.Option.IF_METAGENERATION_MATCH),
+        composeObjectReqBuilder::setIfMetagenerationMatch);
+    ifNonNull(
+        (String) optionsMap.get(StorageRpc.Option.PREDEFINED_ACL),
+        composeObjectReqBuilder::setDestinationPredefinedAcl);
+    ifNonNull(
+        (String) optionsMap.get(StorageRpc.Option.KMS_KEY_NAME),
+        composeObjectReqBuilder::setKmsKey);
```

Comment:
I think we also need to account for encryption options here, similar to https://github.com/googleapis/java-storage/blob/dc78526343c30ea85fd05c6977696309d687d958/google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java#L1181-L1184